### ISSUE TITLE
require-atomic-updates: document.title (false positive)

### ISSUE THREAD

**Issue by Lonniebiz** [eslint/eslint#11967]
I'm using eslint 6.0.1 and believe I'm getting a false positive for the require-atomic-updates rule. Here's a code sample which reproduces the false-positive:
```
async function test()
{
	document.createElement("div");
	await Promise.resolve();
	document.title = "Test";
}
test();
```
The line `document.title = "Test";` (above) produces this error:
```
Possible race condition: `document.title` might be reassigned based 
on an outdated value of `document.title`. (require-atomic-updates)
```

However, if I change `document.title` to `window.document.title` eslint does **NOT** produce the error:
```
async function test()
{
	document.createElement("div");
	await Promise.resolve();
	window.document.title = "Test";
}
test();
```

**Comment by ljharb** [comment]
This JS would cause the problem being warned about:
```js
document.title = 'original title';
test();
document.title = 'surprise!';
```
In other words, the value you're clobbering (`document.title`) wouldn't be "original title", it would be "surprise!" - that's a potential source of bugs.

**Comment by Lonniebiz** [comment]
@ljharb : `document.title` is just node in the DOM. Obviously, it can be changed by any function that has access to the DOM later in other functions (by designed). Are you advocating that estlint should error anytime a function modifies a DOM node or any variable outside of that function (even when there is no other instance of that value being modified in the entire rest of the program)?

Of course not! Therefore, this is a bug.

Why doesn't eslint **also** produce a race condition error when `document.title` is changed to `window.document.title`? Would it no less be the same problem you're describing?

For example, esilint does **NOT** produce the race condition error:
```
async function test()
{
	document.createElement("div");
	await Promise.resolve();
	window.document.title = "Test";
}
test();
```

Also, if you leave it with `document.title` instead of `window.document.title` (as original), but remove the `document.createElement("div")` line, this also does **NOT** produce race condition error :
```
async function test()
{
	await Promise.resolve();
	document.title = "Test";
}
test();
```

Keep in mind, that the entire program **is** only those 7 lines you see in the first code segment of my original post. Where in those 7 lines do you see two things racing to set a value?

I agree that the function I posted is not pure (from a functional programming perspective), but there is no race condition. And, no error should occur. And, this is a bug.

**Comment by platinumazure** [comment]
I have to agree, I'm not sure why this is anything but a bug. `document.title` is never read, and the reference to `document` is never modified. And the value assigned to `document.title` is a constant. If `document.title` were read anywhere, then we have a race condition for sure. But I could be missing something.

As for `document` vs `window.document`: In general, ESLint is not aware of the global object (`window` in browsers, `global` in Node, etc.) due to the need to be runtime-agnostic. So I wouldn't read too much into that if I were you.

**Comment by mysticatea** [comment]
Thank you for your report.

I think this is a duplicate of #11899. Please track that issue.
